helix-core/src/normalized_url.rs

```diff
@@ -0,0 +1,320 @@
+//! [`NormalizedUrl`]s are used to make comparisons more logical for URLs, notably those holding
+//! Windows paths.
+use std::borrow::Cow;
+use std::cmp::Ordering;
+use std::fmt;
+use std::path::{Path, PathBuf};
+
+use url::Url;
+
+/// Normalize a [`Url`] to make it comparable across idiosyncracies from different tools.
+///
+/// # Example case
+///
+/// Helix gets paths from other sources, like LSPs. In such a case, Helix can't know in advance for
+/// every source what the case will be for the prefix component on Windows (`C:` or `c:`). Since
+/// this can hinder comparisons, especially in [`Url`]s, this structure gives access to the opposite
+/// casing of the `Url` and uses it to compare.
+#[derive(Debug, Clone)]
+pub struct NormalizedUrl {
+    base: Url,
+    /// When appropriate, this contains the base URL with the drive letter in the opposite case.
+    ///
+    /// `file://C:/my/windows/path` in `base` means this is `Some(file://c:/my/windows/path)`.
+    windows_normalized: Option<Url>,
```

Comment:
I still think that this should just be
```rust
pub struct NormalizedUrl {
    base: Url
}
```
Then all the important comparison traits should be deriveable. Downside is that you will always have to convert both URLs to `NormalizedUrl`s when comparing.

I can see pros and cons for both. Doing what is done here is a bit easier at the comparison site, with the tradeoff of the struct being twice the size and complicating the implementation slightly. The other way is less ergonomic, but the implementation is simpler and the size of the struct will be the same as `Url`.

Reply:
I don't think size is much of a concern right now, here are the point of use of `NormalizedUrl`:

- `helix_lsp::client::Client`, `root_uri: Option<NormalizedUrl>`. I removed it because it was only used to be sent immediately to the LSP server, no need to construct a normalization only to fail to use it after that. *bonus, it also allowed me to remove `clone` calls*
- `helix_lsp::client::Client::text_document_did_open`, `uri` parameter: I removed this one since it's directly sent to the server, made the helix code pass the owned URL instead of a reference to the NormalizedUrl. With `NormalizedUrl::into_base()` that allowed me to remove a useless clone too :tada:
- `helix_term::commands::lsp`, `impl ui::menu::Item for lsp::SymbolInformation`: here it's needed since it's comparing a Helix URL to an LSP one, precisely the use case it was written for. This function is called once for each time a menu is opened with a symbol to show, which means the comparison can easily be done hundreds of time. Speeding it up means more UI responsiveness.
- `helix_term::commands::lsp::PickerDiagnostic.url`, used for comparison but also stored, measuring would be needed to know which costs more, see `sym_picker()` and `diag_picker()` below
- `helix_term::commands::lsp::sym_picker()`, comparison for the path of each symbol, here it's very probably more useful to store the normalization than recompute it for each symbol
- `helix_term::commands::lsp::diag_picker()` (twice !) In the `BTreeMap` it will inflate the size but also allow much more efficient comparison when looking for an item. Hard to say what's more costly. The other use is the same as `sym_picker()`
- `helix_view::document::Document::url()`, allocates each time but it's not called that much, mostly on opening a file picker (where comparison is done way more) and communicating using `Document::(versioned_)identifier()`, where allocating is done only to cast back into base immediately so initialization is done more. Launching the language server too, but that's a one time cost so one way or another it won't affect performances.
- `helix_view::editor::Editor.diagnostics`: another `BTreeMap`, where I think storage is done more than comparison, but really that would depend on the use of the diagnostic picker. I use it very often so a single batch of diagnostics will usually end up being compared 5 to 10 times even if there are only a few of them, but other people may have other usage pattern. Hard to tell really, without backing data.

Overall I think the current way is good, but the 'normalize during comparison only' way would be good too. Without backing data, I will say that allocating once and then never again, especially when the cost windows normalization won't happen at all for the people on Unix system, is better for allocators than maybe allocating at each comparison.

At least reviewing the usage points helped me find a few improvements ! Thanks for sticking with me on this PR :)